### PR TITLE
COCOS-24 - Add events service via http

### LINKED ISSUE #24: Implement notifications mechanism
Manager and Agent need to notify Computation service of the status

### PR DESCRIPTION
**Pull request by SammyOina**
Adds events service that sends notifications via HTTP.
Relates to https://github.com/ultravioletrs/prism/pull/54
Resolves #24

### CODE REVIEW

**Review by drasko — APPROVED** [review]
LGTM

File: agent/README.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ The service is configured using the environment variables from the following tab
 | AGENT_GRPC_PORT          | Agent service gRPC port                                | 7002                           |
 | AGENT_GRPC_SERVER_CERT   | Path to gRPC server certificate in pem format          | ""                             |
 | AGENT_GRPC_SERVER_KEY    | Path to gRPC server key in pem format                  | ""                             |
+| COCOS_NOTIFICATION_SERVER_URL | Server to receive notification events from agent.      | http:/localhost:9000           |
```
Comment (drasko) [inline agent/README.md:20]: Please align

File: agent/state.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 	mglog "github.com/absmach/magistrala/logger"
 )
 
+//go:generate stringer -type=state
```
Comment (drasko) [inline agent/state.go:13]: What's this?

File: manager/README.md [hunk 3]
```diff
@@ -19,6 +19,7 @@ The service is configured using the environment variables from the following tab
 | MANAGER_GRPC_SERVER_KEY  | Path to server key in pem format                         |                                   |
 | COCOS_JAEGER_URL         | Jaeger server URL                                        | http://localhost:14268/api/traces |
 | MANAGER_INSTANCE_ID      | Manager service instance ID                              |                                   |
+| COCOS_NOTIFICATION_SERVER_URL | Server to receive notification events from agent.        | http:/localhost:9000              |
```
Comment (drasko) [inline manager/README.md:22]: Align

File: manager/service.go [hunk 4]
```diff
@@ -62,7 +81,19 @@ func (ms *managerService) Run(ctx context.Context, c *Computation) error {
...
+		}
+		detailB, merr := json.Marshal(detail)
+		if merr != nil {
+			ms.logger.Warn(merr.Error())
+			return err
+		}
+		ms.publishEvent("vm-provision", c.Id, "failed", detailB)
 		return err
 	}
 	// different VM guests can't forward ports to the same ports on the same host
```
Comment (drasko) [inline manager/service.go:99]: Capitalize comments

**Review by drasko — CHANGES_REQUESTED** [review]

File: agent/state.go (hunk 2 above)
Comment (SammyOina) [reply agent/state.go:13]: states are defined as `int` we use go generate stringer to generate string of the constants for submission of events https://pkg.go.dev/golang.org/x/tools/cmd/stringer

File: agent/service.go [hunk 5]
```diff
@@ -204,25 +207,41 @@ func (as *agentService) Attestation(ctx context.Context) ([]byte, error) {
...
 	var cancel context.CancelFunc
 	if as.computation.Timeout.Duration != 0 {
 		ctx, cancel = context.WithDeadline(ctx, time.Now().Add(as.computation.Timeout.Duration))
 		defer cancel()
 	}
+	as.publishEvent("in-progress", json.RawMessage{})()
 	result, err := run(ctx, as.algorithms[0], as.datasets[0])
 	if err != nil {
 		as.runError = err
+		detail := struct {
```
Comment (drasko) [inline agent/service.go:223]: What is this that is called `detail`? It is not a good and self-describing name.

**Review by drasko — CHANGES_REQUESTED** [review]

File: agent/service.go (hunk 5 above)
Comment (SammyOina) [reply agent/service.go:223]: Details field is an instructed field allowing to add event details as raw json

File: manager/service.go [hunk 6]
```diff
@@ -52,6 +60,17 @@ func (ms *managerService) Run(ctx context.Context, c *Computation) error {
...
 	dur, err := time.ParseDuration(c.Timeout)
 	if err != nil {
+		detail := struct {
+			Error string `json:"error"`
+		}{
+			Error: err.Error(),
+		}
+		detailB, merr := json.Marshal(detail)
+		if merr != nil {
+			ms.logger.Warn(merr.Error())
```
Comment (dborovcanin) [inline manager/service.go:70]: We avoid adding logger as a dependency to the svc, i.e. we separate (as much as possible) logging from the business logic.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
There is no need to check for the error in computation Timeout at the moment. We will probably just remove timeout and leave control over computation to be completely manual; eventually we might add some timeout as an extra-safety measurement.

**Review by drasko — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**